lighthouse-core/gather/gatherers/console.js

```diff
@@ -0,0 +1,118 @@
+/**
+ * @license Copyright 2020 The Lighthouse Authors. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+
+/**
+ * @fileoverview Gathers console deprecation and intervention warnings logged by Chrome.
+ */
+
+'use strict';
+
+const Gatherer = require('./gatherer.js');
+
+class Console extends Gatherer {
+  constructor() {
+    super();
+    /** @type {LH.Artifacts.ConsoleMessage[]} */
+    this._logEntries = [];
+
+    this._onConsoleAPICalled = this.onConsoleAPICalled.bind(this);
+    this._onExceptionThrown = this.onExceptionThrown.bind(this);
+    this._onLogEntryAdded = this.onLogEntry.bind(this);
+  }
+
+  /**
+   * @param {LH.Crdp.Runtime.ConsoleAPICalledEvent} event
+   */
+  onConsoleAPICalled(event) {
+    const level = event.type;
+    if (level !== 'warning' && level !== 'error') {
+      // Only gather warnings and errors for brevity.
+      return;
+    }
+    const args = event.args || [];
+    const text = args.map(a => a.value || a.description || '').filter(Boolean).join(' ');
+    if (!text) {
+      return;
+    }
+    const {url, lineNumber, columnNumber} =
+      event.stackTrace && event.stackTrace.callFrames[0] || {};
+    /** @type {LH.Artifacts.ConsoleMessage} */
+    const consoleMessage = {
+      source: level === 'warning' ? 'console.warn' : 'console.error',
+      level,
+      text,
+      stackTrace: event.stackTrace,
+      timestamp: event.timestamp,
+      url,
+      lineNumber,
+      columnNumber,
+    };
+    this._logEntries.push(consoleMessage);
+  }
+
+  /**
+   * @param {LH.Crdp.Runtime.ExceptionThrownEvent} event
+   */
+  onExceptionThrown(event) {
+    const text = event.exceptionDetails.exception ?
+          event.exceptionDetails.exception.description : event.exceptionDetails.text;
+    if (!text) {
+      return;
+    }
+    /** @type {LH.Artifacts.ConsoleMessage} */
+    const consoleMessage = {
+      source: 'exception',
+      level: 'error',
+      text,
+      stackTrace: event.exceptionDetails.stackTrace,
+      timestamp: event.timestamp,
+      url: event.exceptionDetails.url,
+      lineNumber: event.exceptionDetails.lineNumber,
+      columnNumber: event.exceptionDetails.columnNumber,
+    };
+    this._logEntries.push(consoleMessage);
+  }
+
+  /**
+   * @param {LH.Crdp.Log.EntryAddedEvent} event
```

Comment:
these are the deprecations/warnings what used to be `ConsoleMessages` right?

Reply:
Right